File: Baekjoon/Class3/38.14940 쉬운 최단거리.py
from collections import deque
import sys
input = sys.stdin.readline

MOVE_DIR = ((-1, 0), (1, 0), (0, -1), (0, 1))

n, m = map(int, input().split())

graph = [list(map(int, input().split())) for _ in range(n)] #격자 그래프
depths = [[0] * m for _ in range(n)]
bfs = deque()
for r, row in enumerate(graph):
    for c, num in enumerate(row):
        if num == 1:
            depths[r][c] = -1
        if num == 2: #목표 지점
            bfs.append((r, c))

while bfs: #너비 우선 탐색
    start_r, start_c = bfs.popleft()
    for move_r, move_c in MOVE_DIR:
        end_r = start_r + move_r
        end_c = start_c + move_c
        if end_r < 0 or end_r >= n or end_c < 0 or end_c >= m:
            continue
        if graph[end_r][end_c] == 1: #갈 수 있음
            graph[end_r][end_c] = 0 #방문 표시
            depths[end_r][end_c] = depths[start_r][start_c] + 1 #깊이 계산
            bfs.append((end_r, end_c))

print("\n".join(" ".join(map(str, depths[i])) for i in range(n)))

File: Baekjoon/Class3/38.14940 쉬운 최단거리.js
const inputFile = process.platform === "linux" ? "/dev/stdin" : "./input.txt";
const input = require("fs").readFileSync(inputFile, "utf8").trim().split("\n");

const MOVE_DIR = [[-1, 0], [1, 0], [0, -1], [0, 1]];

const [n, m] = input[0].split(" ").map(Number);

const graph = []; // 격자 그래프
for(let i = 1; i <= n; i++)
    graph.push(input[i].split(" ").map(Number));

const depths = Array.from({ length : n }, () => new Array(m).fill(0));
let bfsHead = 0;
const bfs = [];
for(let r = 0; r < n; r++)
    for(let c = 0; c < m; c++) {
        const num = graph[r][c]
        if(num === 1)
            depths[r][c] = -1;
        else if(num === 2) // 목표 지점
            bfs.push([r, c]);
    }

while(bfsHead < bfs.length) { // 너비 우선 탐색
    const [startR, startC] = bfs[bfsHead++];
    for(const [moveR, moveC] of MOVE_DIR) {
        const endR = startR + moveR;
        const endC = startC + moveC;
        if(endR < 0 || endR >= n || endC < 0 || endC >= m)
            continue;
        if(graph[endR][endC] === 1) { // 갈 수 있음
            graph[endR][endC] = 0; // 방문 표시
            depths[endR][endC] = depths[startR][startC] + 1; // 깊이 계산
            bfs.push([endR, endC]);
        }
    }
}

console.log(depths.map((row) => row.join(" ")).join("\n"));